firewalld resource: inspect the status and configuration of firewalld

dromazmj opened the pull request:
Signed by: dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
New Resource - firewalld

Signed-off-by: dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
New Resource - firewalld

Signed-off-by: dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
New Resource - firewalld

Signed-off-by: dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
Merge branch 'master' into al/firewalld

dromazmj committed:
New Resource - firewalld

Signed-off-by: dromazos <dromazmj@dukes.jmu.edu>

adamleff reviewed (changes requested):
Please apply feedback from #2064 and others to this PR and let me know when it's ready for re-review.

dromazmj committed:
Modifications to new resourec - firewalld

Signed-off-by: dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
Modifications to new resource - firewalld

Signed-off-by: dromazos <dromazmj@dukes.jmu.edu>

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -0,0 +1,88 @@
+
+Use the `firewalld` InSpec audit resource to test rules set to match IP address's with hostnames.
+
+An firewalld rule specifies an IP address and what it's hostname is along with an aliase names it can have.
```
`s/it's/its` and `s/aliase/alias`. Also, this looks like the description for the `etc_hosts` resource, not the `firewalld` resource.

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -0,0 +1,88 @@
+
+The `active_zones` property displays the open zone and all its open interfaces
+
+    its ('active_zones') { should eq "public\n  interfaces: enp0s3\n" }
```
This seems like a bad user experience to me. A user shouldn't have to worry about including additional output when they're querying zones. `active_zones` should just return an array of names of zones that are active, right?

If we need to, we can always create a FilterTable of data for the zones... zone name, interface list, and state (active/inactive).

dromazmj replied:
I understand that, Yeah this is a good place for a FilterTable so the user can query for interfaces by zones.

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -0,0 +1,88 @@
+
+`have_rule_enabled` returns true or false if the rich-rule has been specified in the zone.
+
+    it { should have_rule_enabled('rule family=ipv4 source address=192.168.0.14 accept', 'public') }
```
Can we strip `rule ` from the string we match on? That's really a keyword for firewalld to know this is a rule, but the actual rule itself is all the rest of the stuff.

dromazmj replied:
Sure thing

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -0,0 +1,88 @@
+
+## Examples
+
+### Check the services allowed in a zone. `services_bound` returns a string list of services allowed through the zone.
```
The example doesn't include a way for the user to specify the zone like the description says. How would the user specify that?

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -0,0 +1,88 @@
+      its ('services_bound')  { should eq ['ssh', 'icmp'] }
+    end
+
+### Check the sources allowed in a zone. `sources_bound` returns a string list of sources allowed through the zone.
```
The example doesn't include a way for the user to specify the zone like the description says. How would the user specify that?

adamleff reviewed (changes requested):
@dromazmj before I dig into the code, I'd like to agree on the documentation / command structure. I think it will help us figure out how the code might need to change, whether we have a need for FilterTable, etc. Thanks!

dromazmj committed:
Modifications to new resource firewalld

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -4,34 +4,64 @@ title: About the firewalld Resource
+The `sources` property is used in conjunction with the where class to display open sources in an active zone.
+
+    describe firewalld.where { zone == 'public' } do
+      its('souces') { should cmp ['192.168.1.0/24', '192.168.1.2'] }
```
`s/souces/sources`

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -4,34 +4,64 @@ title: About the firewalld Resource
-    its ('default_zone') { should eq 'public' }
+    its('default_zone') { should eq 'public' }
 
 ### default_zone
```
`s/default_zone/active_zones`

adamleff commented on docs/resources/firewalld.md.erb:
```diff
@@ -4,34 +4,64 @@ title: About the firewalld Resource
 The `active_zones` property displays the open zone and all its open interfaces
 
-    its ('active_zones') { should eq "public\n  interfaces: enp0s3\n" }
+    its('active_zones') { should eq "public\n  interfaces: enp0s3\n" }
```
Is this a documentation error, or is this still expected behavior? I believe there was a prior comment from me that we shouldn't expect the user to format the output like this in order to check for active zones. It should only include the zone, no newlines or interfaces, etc.

adamleff commented on test/unit/resources/firewalld_test.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  it 'verify firewalld detects a whether or not a service is allowed in a zone' do
+    resource.stubs(:has_service_enabled_in_zone?).with('public', 'ssh').returns(true)
```
Lines 32, 36, 40, 44, and 48 aren't actually testing anything - they're simply stubbing out a response for a test to be executed at some point. Please add tests for each of these.

adamleff commented on test/unit/resources/firewalld_test.rb:
```diff
@@ -0,0 +1,50 @@
+  it 'verify firewalld detects a whether or not a rule is enabled in a zone' do
+    resource.stubs(:has_rule_enabled?).with('public', 'rule family=ipv4 source address=192.168.0.14 accept').returns(true)
+  end
+end
```
Missing a test for the `active_zones` method?

adamleff commented on test/unit/resources/firewalld_test.rb:
```diff
@@ -0,0 +1,50 @@
+  centResource = MockLoader.new(:centos7).load_resource('firewalld')
+
+  it 'verify firewalld detects a zone' do
+    centResource.stubs(:has_zone?).with('public').returns(true)
```
This line, and line 13, are not actual tests but rather stubs/mocks. Please add tests.

dromazmj committed:
Modifications to new resource - firewalld

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,144 @@
+      firewalld_command("firewall-cmd --zone=#{query_zone} --list-services").split(' ')
+    end
+
+    def firewalld_command(command)
```
I think we discussed this last week, but just in case it slipped through the cracks...

I'd change this resource to look something like this:

```ruby
def firewalld_command(command_args)
  command_to_run = "firewall-cmd #{command_args}"
  result = inspec.command(command_to_run)
  # rest of method here...
end
```

... and then you don't have to repeat `firewall-cmd` in every call to it.  For example:

```ruby
firewalld_command("--zone=#{query_zone} --list-sources").split(' ')
```

This just eliminates a ton of repetition. Not a huge deal, but something I think makes sense to clean up.

dromazmj committed:
Changing firewalld_command method to prepend the command with 'firewall-cmd' to reduce code reuse.

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+      return false unless installed?
+      result = firewalld_command('--state')
+      return result unless result == "running\n" || result == "not running\n"
+      result[0, result.length-1] == 'running'
```
I'm not entirely sure what you're trying to accomplish with this method. Since it's a "question-mark method" it should return `true` or `false`... but in line 58 you could potentially return a completely different result string.

Can this whole method just be trimmed down to:

```ruby
!!result.match(/^running/)
```

That will return `true` if the result starts with `running` -- otherwise, return false.

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+      if result.stderr != ''
+        return "Error on command #{command}: #{result.stderr}"
+      end
+      result.stdout
```
See all the above comments re: newlines..  I'm pretty sure `firewall-cmd` ends all stdout with a newline character.

```ruby
result.stdout.strip
```

... is probably what you want to do so no caller needs to worry about trailing newline characters or spaces.

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+    def default_zone
+      # return: word associated with the name of the default zone
+      # example: 'public'
+      firewalld_command('--get-default-zone')[0..-2]
```
This only works if the command always ends with a newline character, and it's a bit hacky to do it this way.

The `firewalld_command` should return output that doesn't have a newline character, and then this (and other methods similar to this) can not have to worry about doing any string slicing like this.

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+    end
+
+    def has_service_enabled_in_zone?(query_service, query_zone = default_zone)
+      firewalld_command("--zone=#{query_zone} --query-service=#{query_service}")[0..-2] == 'yes'
```
See above comment - let's do what's necessary to get rid of the string slicing and instead just match on the output so this can become something like:

```ruby
firewalld_command("--zone=#{query_zone} --query-service=#{query_service}") == 'yes'
```

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+    def service_ports_enabled_in_zone(query_service, query_zone = default_zone)
+      # return: String of ports open
+      # example: ['22/tcp', '4722/tcp']
+      firewalld_command("--zone=#{query_zone} --service=#{query_service} --get-ports --permanent").split(' ')
```
Re: the above comments, splitting like this without ensuring the line doesn't end in a newline will mean the last item in the array will contain the newline itself. Let's modify firewalld_command to ensure that doesn't happen.

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+    end
+
+    def has_port_enabled_in_zone?(query_port, query_zone = default_zone)
+      firewalld_command("--zone=#{query_zone} --query-port=#{query_port}")[0..-2] == 'yes'
```
Same re: string slicing.

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,145 @@
+
+    def has_rule_enabled?(rule, query_zone = default_zone)
+      rule = 'rule ' + rule
+      firewalld_command("--zone=#{query_zone} --query-rich-rule=#{rule}")[0..-2] == 'yes'
```
Same re: string slicing.

adamleff reviewed (changes requested):
@dromazmj this is in pretty good shape. There is some code cleanup we need to do, and we need to enhance our tests with multiple zones to ensure your resource handles multiple zones properly.

dromazmj committed:
Modifications made
	* installed? method now tells by checking if firewall-cmd is a command on the system
	* The firewalld_command method now strips the stdout of the return
	* added another test for testing multiple active zones

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

dromazmj committed:
Fixing rake lint issue

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

dromazmj committed:
Fixing match and returning boolean for seeing if firewalld is running

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

dromazmj committed:
Fixing lint issues

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

adamleff commented on lib/resources/firewalld.rb:
```diff
@@ -0,0 +1,144 @@
+    def running?
+      return false unless installed?
+      result = firewalld_command('--state')
+      result =~ /^running/ ? true : false
```
Sorry about Rubocop being a pain on this.

Another option that would make Rubocop happy would be:

```ruby
def running?
  return false unless installed?
  !firewalld_command('--state').match(/^running/).nil?
end
```

I'm fine leaving this as-is though if you are.

adamleff reviewed (approved):
LGTM @dromazmj!

dromazmj committed:
Merge branch 'master' into al/firewalld

dromazmj committed:
Empty commit to rerun.  Accidentally updated branch.

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

dromazmj committed:
Merge branch 'al/firewalld' of https://github.com/aaronlippold/inspec into al/firewalld

dromazmj committed:
Rerunning test, accidentally updated branch. needs sign off commit

Signed-off-by: dromazmj <dromazmj@dukes.jmu.edu>

arlimus reviewed (approved):
Thank you @dromazmj this is very much appreciated!! 👍 

Also kudos @adamleff for guiding this through 😁
